extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeDataProvider.ts

```diff
@@ -0,0 +1,147 @@
+
+export class ControllerTreeDataProvider implements vscode.TreeDataProvider<TreeNode>, IControllerTreeChangeHandler {
+
+	private _onDidChangeTreeData: vscode.EventEmitter<TreeNode> = new vscode.EventEmitter<TreeNode>();
+	public readonly onDidChangeTreeData: vscode.Event<TreeNode> = this._onDidChangeTreeData.event;
+	private root: ControllerRootNode;
+	private credentialProvider: azdata.CredentialProvider;
+
+	constructor() {
+		this.root = new ControllerRootNode({ treeChangeHandler: this });
+		this.loadSavedControllers();
+	}
+
+	public async getChildren(element?: TreeNode): Promise<TreeNode[]> {
+		if (element) {
+			return element.getChildren();
+		}
+
+		if (this.root.hasChildren) {
+			return this.root.getChildren();
+		} else {
+			return [new AddControllerNode()];
+		}
+	}
+
+	public getTreeItem(element: TreeNode): vscode.TreeItem | Thenable<vscode.TreeItem> {
+		return element.getTreeItem();
+	}
+
+	public addController(
+		url: string,
+		username: string,
+		password: string,
+		rememberPassword: boolean,
+		masterInstance?: IEndPoint
+	): void {
+		this.root.addControllerNode(url, username, password, rememberPassword, masterInstance);
+		this.notifyNodeChanged();
+	}
+
+	public deleteController(url: string, username: string): ControllerNode {
+		let deleted = this.root.deleteControllerNode(url, username);
+		if (deleted) {
+			this.notifyNodeChanged();
+		}
+		return deleted;
+	}
+
+	public notifyNodeChanged(node?: TreeNode): void {
+		this._onDidChangeTreeData.fire(node);
+	}
+
+	public async loadSavedControllers(): Promise<void> {
+		let config = vscode.workspace.getConfiguration(ConfigNamespace);
+		if (config && config.controllers) {
+			let controllers = config.controllers;
+			this.root.clearChildren();
+			for (let c of controllers) {
+				let password = undefined;
+				if (c.rememberPassword) {
+					password = await this.getPassword(c.url, c.username);
+				}
+				this.root.addChild(new ControllerNode({
+					url: c.url,
+					username: c.username,
+					password: password,
+					rememberPassword: c.rememberPassword,
+					parent: this.root,
+					treeChangeHandler: this
+				}));
+			}
+			this.notifyNodeChanged();
+		}
+	}
+
+	public async saveControllers(): Promise<void> {
+		let controllers = this.root.children.map(e => {
+			let controller = e as ControllerNode;
+			return {
+				url: controller.url,
+				username: controller.username,
+				password: controller.password,
+				rememberPassword: !!controller.rememberPassword
+			};
+		});
+
+		let controllersWithoutPassword = controllers.map(e => {
+			return {
+				url: e.url,
+				username: e.username,
+				rememberPassword: e.rememberPassword
+			};
+		});
+
+		try {
+			await vscode.workspace.getConfiguration(ConfigNamespace).update('controllers', controllersWithoutPassword, true);
+		} catch (error) {
+			vscode.window.showErrorMessage(error.message);
+		}
+
+		for (let e of controllers) {
+			await this.savePassword(e.url, e.username, e.password);
```

Comment:
We should only do this if `rememberPassword` is true. Otherwise we save all passwords (just don't look them up after).

Reply:
fixed.